smarthome.go

```diff
@@ -67,15 +69,66 @@ func (s *SmartHome) Handle() http.HandlerFunc {
 		defer r.Body.Close()
 
 		res := s.decodeAndHandle(oauth.GetAgentUserIdFromHeader(r), r.Body)
-		if err := json.NewEncoder(io.MultiWriter(w, log.Writer())).Encode(res); err != nil {
-			log.Println("error encoding response:", err)
+		if err := json.NewEncoder(w).Encode(res); err != nil {
+			if s.Log != nil {
+				s.Log.Println("error encoding response:", err)
+			}
 			http.Error(w, "error encoding response: "+err.Error(), http.StatusInternalServerError)
 		}
 	}
 }
 
+func (s *SmartHome) executeCommandForResponse(id string, dev Device, ex proto.CommandRequest) proto.CommandResponse {
+	response := proto.CommandResponse{
+		Ids:       []string{id},
+		ErrorCode: proto.ErrorCodeNotSupported.Error(),
+		//States:    make(map[string]interface{}),
+	}
+	// response.States["online"] = true
+
+	// check all traits...
+	for _, trait := range dev.DeviceTraits() {
+		// for all the commands...
+		for _, cmd := range trait.TraitCommands() {
+			// for the right command
+			if ex.Command == cmd.Name() {
+				// and execute
+				ctx := Context{Target: dev}
+				if err := cmd.Execute(ctx, ex.Params); err == nil {
+					response.Status = proto.CommandStatusSuccess
+					response.ErrorCode = ""
+
+					// States are optional in this response
```

Comment:
Does the response time dramaticly increase if you leave this in?

Reply:
unknown, it depends on what the user has defined in the state handlers, for example with insteon it means doing a HTTP request, and their API responds slowly